Models

MaskaSaw opened the pull request:
Пулл реквест на два таска: на добавление и конфигурацию моделей, контроллеров и их методов и на удаление лишних файлов из репозитория.

MaskaSaw committed:
Create models

MaskaSaw committed:
Create .gitignore

MaskaSaw committed:
Controllers signature update

Changed routings, signature of actions in RoomsController class

MaskaSaw committed:
Comment changes

Minor comment changes

MaskaSaw committed:
Setup actions of all controllers

Added some actions for controllers, changed routings

MaskaSaw committed:
Remove unnecessary files

Removed all unnecessary files according to .gitignore

MaskaSaw committed:
Remove unnecessary files

Removed unnecessary files from Angular client and admin's app

MaskaSaw committed:
Minor fixes for models

Changed models members, returned originally generated properties.

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/HotelsController.cs:
```diff
@@ -0,0 +1,129 @@
+        }
+
+        // PUT: api/Hotels/5
+        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
```
Do you really need this comment here? What value does it add to your code?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/HotelsController.cs:
```diff
@@ -0,0 +1,129 @@
+        }
+
+        // POST: api/Hotels
+        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
```
Do you really need this comment here? What value does it add to your code?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/HotelsController.cs:
```diff
@@ -0,0 +1,129 @@
+                {
+                    throw;
+                }
+            }
```
In what circumstances will this situation be possible?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/HotelsController.cs:
```diff
@@ -0,0 +1,129 @@
+
+            _context.Hotels.Remove(hotel);
+            var rooms = await _context.Rooms.ToListAsync();
+            rooms = rooms.Where(x => x.HotelId == id).ToList();
```
In case when several method calls are chained start each method call from a new line.
```suggestion
            rooms = rooms
                 .Where(x => x.HotelId == id)
                 .ToList();
```

andrei-kruhliakou-itechart replied:
Why do you need two ToList() calls? How will it translate to DB calls?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/ReservationsController.cs:
```diff
@@ -0,0 +1,115 @@
+        public async Task<ActionResult<IEnumerable<Reservation>>> GetReservations(int userId)
+        {
+            var reservations = await _context.Reservations.ToListAsync();
+            return reservations.Where(x => x.UserId == userId).ToList();
```
Why calling ToList() twice? How will it translate to SQL queries?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/ReservationsController.cs:
```diff
@@ -0,0 +1,115 @@
+        }
+
+        // PUT: api/Reservations/5
+        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
```
Is this really needed?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/ReservationsController.cs:
```diff
@@ -0,0 +1,115 @@
+        }
+
+        // POST: api/Reservations
+        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
```
Is this really needed?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Hotels.csproj:
```diff
@@ -5,7 +5,19 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="5.0.2" />
```
Why do you need SQLite?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/Hotel.cs:
```diff
@@ -0,0 +1,27 @@
+    {
+        public Hotel()
+        {
+            //Rooms = new HashSet<Room>();
```
Do not leave commented out code

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/HotelsDBContext.cs:
```diff
@@ -0,0 +1,118 @@
+        {
+            if (!optionsBuilder.IsConfigured)
+            {
+                optionsBuilder.UseSqlServer("Server=wsa-033-71;Database=HotelsDB;Trusted_Connection=True;");
```
Is it a good idea to commit you computer name to public repository?
Also what if I want to use the database on some other machine?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/HotelsDBContext.cs:
```diff
@@ -0,0 +1,118 @@
+
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            modelBuilder.HasAnnotation("Relational:Collation", "Cyrillic_General_CI_AS");
```
Why do you need this?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/HotelsDBContext.cs:
```diff
@@ -0,0 +1,118 @@
+            {
+                entity.ToTable("Hotel");
+
+                entity.Property(e => e.Id).ValueGeneratedNever();
```
Here and below in this file - put second and following calls on a new line.

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/HotelsDBContext.cs:
```diff
@@ -0,0 +1,118 @@
+                entity.HasOne(d => d.Room)
+                    .WithMany(p => p.Reservations)
+                    .HasForeignKey(d => d.RoomId)
+                    .HasConstraintName("FK__Reservati__RoomI__32E0915F");
```
Is it a good name? Do you really need to have named constraint?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/appsettings.json:
```diff
@@ -1,4 +1,7 @@
 {
+  "ConnectionStrings": {
+    "DefaultConnection": "Server=wsa-033-71;Database=HotelsDB;Trusted_Connection=True;MultipleActiveResultSets=true"
```
Is it a good idea to put your computer name in public repo?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/Room.cs:
```diff
@@ -0,0 +1,28 @@
+    {
+        public Room()
+        {
+            //Reservations = new HashSet<Reservation>();
```
Do not leave commented out code

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/ReservationsController.cs:
```diff
@@ -0,0 +1,115 @@
+        [HttpPut("{id}")]
+        public async Task<IActionResult> PutReservation(int id, Reservation reservation)
+        {
+            if (id != reservation.Id)
```
What if I send you a reservation data, that contains wrong roomId? Or userId? Valid userId of another user?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/ReservationsController.cs:
```diff
@@ -0,0 +1,115 @@
+        [HttpPost]
+        public async Task<ActionResult<Reservation>> PostReservation(Reservation reservation)
+        {
+            _context.Reservations.Add(reservation);
```
Are you sure that reservation object contains valid data?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/RoomsController.cs:
```diff
@@ -0,0 +1,156 @@
+        }
+
+        // GET: api/Hotels/5/Rooms
+        [HttpGet("api/Hotels/{hotelId}/[controller]")]
```
Are you sure this action belong to this controller?
Also why have 2 different actions for getting list of rooms? Do you really need all rooms regardless of a hotel they belong to?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/RoomsController.cs:
```diff
@@ -0,0 +1,156 @@
+            foreach (var room in rooms)
+            {
+                room.Reservations = reservations.Where(x => x.RoomId == room.Id).ToList();
+            }
```
Please see https://docs.microsoft.com/en-us/ef/core/querying/related-data/

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/RoomsController.cs:
```diff
@@ -0,0 +1,156 @@
+                room.Reservations = reservations.Where(x => x.RoomId == room.Id).ToList();
+            }
+        
+            return rooms.Where(x => x.HotelId == hotelId).ToList();
```
Please see https://docs.microsoft.com/en-us/ef/core/querying/client-eval

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/RoomsController.cs:
```diff
@@ -0,0 +1,156 @@
+            }
+
+            var reservations =  await _context.Reservations.ToListAsync();
+            room.Reservations = reservations.Where(x => x.RoomId == id).ToList();
```
Please see https://docs.microsoft.com/en-us/ef/core/querying/related-data/

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/RoomsController.cs:
```diff
@@ -0,0 +1,156 @@
+
+            var reservations = await _context.Reservations.ToListAsync();
+            room.Reservations = reservations.Where(x => x.RoomId == id).ToList();
+            return room;
```
Please see https://docs.microsoft.com/en-us/ef/core/querying/related-data/

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/RoomsController.cs:
```diff
@@ -0,0 +1,156 @@
+
+
+        // PUT: api/Rooms/5
+        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
```
Do you need this comment?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/RoomsController.cs:
```diff
@@ -0,0 +1,156 @@
+        }
+
+        // POST: api/Rooms
+        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
```
Do you need this?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/RoomsController.cs:
```diff
@@ -0,0 +1,156 @@
+        [HttpPost]
+        public async Task<ActionResult<Room>> PostRoom(Room room)
+        {
+            _context.Rooms.Add(room);
```
Are you sure room has valid data? What about hotelId?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/RoomsController.cs:
```diff
@@ -0,0 +1,156 @@
+            _context.Rooms.Remove(room);
+
+            var reservations = await _context.Reservations.ToListAsync();
+            reservations = reservations.Where(x => x.RoomId == id).ToList();
```
Rewrite query

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,113 @@
+        [HttpPost]
+        public async Task<ActionResult<User>> PostUser(User user)
+        {
+            _context.Users.Add(user);
```
Are you sure user object has valid data?
What about passwords, are you going just store them as clear text?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,113 @@
+        }
+
+        // POST: api/Users
+        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
```
Do you need this?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,113 @@
+        }
+
+        // PUT: api/Users/5
+        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
```
Do you need this?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,113 @@
+            }
+
+            _context.Users.Remove(user);
+            var reservations = await _context.Reservations.ToListAsync();
+            reservations = reservations.Where(x => x.UserId == id).ToList();
```
Rewrite

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/Hotel.cs:
```diff
@@ -0,0 +1,27 @@
+﻿using System;
+using System.Collections.Generic;
+
+#nullable disable
```
Why do you need this?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/HotelsDBContext.cs:
```diff
@@ -0,0 +1,118 @@
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Metadata;
+
+#nullable disable
```
Why do you need this?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/HotelsDBContext.cs:
```diff
@@ -0,0 +1,118 @@
+
+            modelBuilder.Entity<Hotel>(entity =>
+            {
+                entity.ToTable("Hotel");
```
Why do you need this?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/HotelsDBContext.cs:
```diff
@@ -0,0 +1,118 @@
+
+            modelBuilder.Entity<Room>(entity =>
+            {
+                entity.ToTable("Room");
```
Why do you need this?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/HotelsDBContext.cs:
```diff
@@ -0,0 +1,118 @@
+
+            modelBuilder.Entity<User>(entity =>
+            {
+                entity.ToTable("User");
```
Why do you need this?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/HotelsDBContext.cs:
```diff
@@ -0,0 +1,118 @@
+                    .IsUnicode(false);
+            });
+
+            OnModelCreatingPartial(modelBuilder);
+        }
+
+        partial void OnModelCreatingPartial(ModelBuilder modelBuilder);
```
Why do you need this?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/Reservation.cs:
```diff
@@ -0,0 +1,23 @@
+﻿using System;
+using System.Collections.Generic;
+
+#nullable disable
```
Why do you need this?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/Room.cs:
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.Collections.Generic;
+
+#nullable disable
```
Why do you need this?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/User.cs:
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+
+#nullable disable
```
Why do you need this?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/User.cs:
```diff
@@ -0,0 +1,22 @@
+    {
+        public User()
+        {
+            //Reservations = new HashSet<Reservation>();
```
Do not leave commented out code

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/Room.cs:
```diff
@@ -0,0 +1,28 @@
+        }
+
+        public int Id { get; set; }
+        public int? RoomNumber { get; set; }
```
My hotel has room numbers like 303A or 405-2. How can I use your product for listing them?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/Room.cs:
```diff
@@ -0,0 +1,28 @@
+        public int? RoomNumber { get; set; }
+        public string RoomType { get; set; }
+        public int? VacantBeds { get; set; }   
+        public int? Cost { get; set; }
```
Why can't I use prices like $49.99?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/Room.cs:
```diff
@@ -0,0 +1,28 @@
+        public string RoomType { get; set; }
+        public int? VacantBeds { get; set; }   
+        public int? Cost { get; set; }
+        public bool? Available { get; set; }
+        public bool? Reserved { get; set; }
```
Why are those nullable? What is the meaning for null in this case?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/Reservation.cs:
```diff
@@ -0,0 +1,23 @@
+    public partial class Reservation
+    {
+        public int Id { get; set; }
+        public int? UserId { get; set; }
+        public int? RoomId { get; set; }
+        public DateTime? StartDate { get; set; }
+        public DateTime? EndDate { get; set; }
+        public TimeSpan? ArrivalTime { get; set; }
+        public bool? Parking { get; set; }
+        public bool? Massage { get; set; }
+        public bool? ExtraTowels { get; set; }
```
Why all those fields are nullable? Does it make sense to have reservation with only Id filled?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/User.cs:
```diff
@@ -0,0 +1,22 @@
+    public partial class User
+    {
+        public User()
+        {
+            //Reservations = new HashSet<Reservation>();
+        }
+
+        public int Id { get; set; }
+        public string Login { get; set; }
+        public string Password { get; set; }
+        public bool? IsAdmin { get; set; }
```
What null value means here?

andrei-kruhliakou-itechart reviewed (changes requested):
How can I see your database schema? How can I deploy it to my computer?

andrei-kruhliakou-itechart commented:
> Пулл реквест на два таска: на добавление и конфигурацию моделей, контроллеров и их методов и на удаление лишних файлов из репозитория.

English please!

MaskaSaw committed:
Fixing models

Changed models structure, reworked database context, added database schema

MaskaSaw committed:
Significant changes in backend

Refactored a lot of everything in controllers, models and DB context

MaskaSaw committed:
Little style refactoring

MaskaSaw committed:
Update Room.cs

MaskaSaw commented:
Tried to fix most of issues; all connected with data validation will be part of one of the next tasks.

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/HotelsController.cs:
```diff
@@ -0,0 +1,114 @@
+            {
+                await _context.SaveChangesAsync();
+            }
+            catch (DbUpdateConcurrencyException)
+            {
+                if (!HotelExists(id))
+                {
+                    return NotFound();
+                }
+                else
+                {
+                    throw;
+                }
+            }
```
What does this code do exactly? Why the hotel might be missing? Is NotFound proper result in this case?
Same question for all controllers

MaskaSaw replied:
It might be missing if in exact the same time it was deleted or modified by the other request, for example. Or am i wrong?

andrei-kruhliakou-itechart replied:
If it was modified by other request why return NotFound?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/HotelsController.cs:
```diff
@@ -0,0 +1,114 @@
+                return NotFound();
+            }
+
+            _context.Hotels.Remove(hotel);
+            var rooms = await _context.Rooms
+                .Where(x => x.HotelId == id)
+                .ToListAsync();
+
+            foreach (var room in rooms) 
+            {
+                _context.Rooms.Remove(room);
+            } 
+
```
What about rooms with reservations?
Also are you sure you want to actually delete data from database? Don't you want to keep history of reservations?
Same question for other controllers

MaskaSaw replied:
I guess we should not keep rooms of deleted hotels in database, because it might provide some issues in the future. For example, it will be possible to make reservation of the room that exists in actually non-existing hotel. There are two possible ways to prevent it: add data validation for every query that will control it and change the model, providing additional field that shows is that room is actually valid or not; or we can just delete all rooms with their hotels from database. It will also delete all reservations connected with such a room.

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/ReservationsController.cs:
```diff
@@ -0,0 +1,114 @@
+            return await _context.Reservations.ToListAsync();
+        }
+
+        // GET: api/User/5/Reservations
+        [HttpGet("api/User/{userId}/Reservations")]
```
Does it really belong here? Also will this route work at all?

MaskaSaw replied:
If we return Reservation data, action that does it should be in Reservation controller. Or am i wrong? Don't know the right answer.
This route works at all.

andrei-kruhliakou-itechart replied:
Judging from the URI it is part of user resource

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/ReservationsController.cs:
```diff
@@ -0,0 +1,114 @@
+
+        // GET: api/Reservations
+        [HttpGet]
+        public async Task<ActionResult<IEnumerable<Reservation>>> GetReservations()
```
What is the use case for getting ALL reservations regardless of hotel or user? 
In general you should not allow API client to get ALL the records from database without limits for any table that can potentially hold a reasonably big number of rows.

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/RoomsController.cs:
```diff
@@ -0,0 +1,116 @@
+                return NotFound();
+            }
+
+            room.Reservations =  await _context.Reservations.Where(x => x.RoomId == id).ToListAsync();
```
Every subsequent method call should be on a new line

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/RoomsController.cs:
```diff
@@ -0,0 +1,116 @@
+
+            _context.Rooms.Remove(room);
+
+            var reservations = await _context.Reservations.Where(x => x.RoomId == id).ToListAsync();
```
Every subsequent method call should be on a new line

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Migrations/20210204085616_InitializingMigration.Designer.cs:
```diff
@@ -0,0 +1,211 @@
+﻿// <auto-generated />
```
The requirements explicitly state that you should use DB first approach with Entity Framework.
It means among other things that you shouldn't use migrations.

MaskaSaw replied:
I used DB first approach, made migrations just to show the database schema

andrei-kruhliakou-itechart replied:
And what are you gonna do when schema changes?

MaskaSaw committed:
Controller fixes

MaskaSaw committed:
Add pagination for reservations

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/ReservationsController.cs:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        // GET: api/Reservations
+        [HttpGet("{page}-{itemsCount}")]
```
Is it a good idea to make page and item count part of the resource name?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/ReservationsController.cs:
```diff
@@ -0,0 +1,117 @@
+        public async Task<ActionResult<IEnumerable<Reservation>>> GetReservations(int page, int itemsCount)
+        {
+            return await _context.Reservations
+                .Skip((page - 1)* itemsCount)
```
What if I pass 2000000 as item count? or 10000000 as page?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/RoomsController.cs:
```diff
@@ -0,0 +1,111 @@
+        [HttpGet]
+        public async Task<ActionResult<IEnumerable<Room>>> GetRooms()
+        {
+            return await _context.Rooms
+                .Include(room => room.Reservations)
+                .ToListAsync();
```
Same question here as for reservations. Why would I need the list of all the rooms regardless of their hotels?

MaskaSaw committed:
Fixing latest issues

Added sql server database schema, fixed controllers, changed connection string, relocated GET methods from Rooms and Reservations controllers.

MaskaSaw committed:
Remove migrations

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/HotelsController.cs:
```diff
@@ -0,0 +1,119 @@
+    public class HotelsController : ControllerBase
+    {
+        private readonly HotelsDBContext _context;
+        private readonly int _itemsCount = 100;
```
Can you think of a better name for this field that would explain its purpose?
Also why don't you use const instead of readonly?

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/HotelsController.cs:
```diff
@@ -0,0 +1,119 @@
+
+        // GET: api/Hotels
+        [HttpGet]
+        public async Task<ActionResult<IEnumerable<Hotel>>> GetHotels([FromQuery] int page)
```
It might be better UX to actually allow user specify number of items he want to get but limit the maximum number of items per page

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/HotelsController.cs:
```diff
@@ -0,0 +1,119 @@
+                }
+                else
+                {
+                    return Conflict(ex);
```
Returned status code is fine now, but why do you pass exception details? Is it OK for user to see them?

andrei-kruhliakou-itechart commented on hotels-backend/HotelsDB/dbo/Tables/Hotel.sql:
```diff
@@ -0,0 +1,15 @@
+    [Image]        NVARCHAR (100) NULL,
+    PRIMARY KEY CLUSTERED ([Id] ASC)
+);
+
+
+
```
Unnecessary empty lines

MaskaSaw committed:
Latest issues fix

andrei-kruhliakou-itechart commented on hotels-backend/HotelsDB/HotelsDB.refactorlog:
```diff
@@ -0,0 +1,10 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
This file should be ignored

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Models/UserDTO.cs:
```diff
@@ -0,0 +1,13 @@
+﻿using System;
```
It is better to place this file in different folder (namespace) to separate DTOs from DB entities.

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/UsersController.cs:
```diff
@@ -86,12 +87,22 @@ public async Task<IActionResult> PutUser(int id, User user)
-        public async Task<ActionResult<User>> PostUser(User user)
+        public async Task<ActionResult<User>> PostUser(UserDTO userDTO)
         {
+            string passwordHash, passwordSalt;
+            CreatePasswordHash(userDTO.Password, out passwordHash, out passwordSalt);
```
Several points here:
1. You can declare out parameters right in the method call
2. Using multiple out parameters is discouraged. Alternatives are tuples or separate class/struct.
3. What are the advantages of using separate method for creating hash and salt? Maybe you can change its scope? Or even consider removing it?

MaskaSaw replied:
I will put this in a separate class when i begin to deal with authorization

andrei-kruhliakou-itechart replied:
Please add todo item in code and/or task in backlog for this.

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/HotelsController.cs:
```diff
@@ -23,11 +23,12 @@ public HotelsController(HotelsDBContext context)
         // GET: api/Hotels
         [HttpGet]
-        public async Task<ActionResult<IEnumerable<Hotel>>> GetHotels([FromQuery] int page)
+        public async Task<ActionResult<IEnumerable<Hotel>>> GetHotels([FromQuery] int page, int requiredNumberOfItems)
```
Parameter name 'requiredNumberOfItems' is quite misleading as we actually want to set number of items **per page**.
How about 'itemsPerPage' instead, should match quite well with _maxItemsPerPage constant.

andrei-kruhliakou-itechart commented on hotels-backend/Hotels/Controllers/HotelsController.cs:
```diff
@@ -14,7 +14,7 @@ namespace Hotels.Controllers
     {
         private readonly HotelsDBContext _context;
-        private readonly int _itemsCount = 100;
+        private const int _maxItemsPerPage = 100;
```
Constants should be named in Pascal case, e.g. MaxItemsPerPage.

MaskaSaw committed:
Controllers and DTO fix
